posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md

```diff
@@ -0,0 +1,613 @@
+linear search for «small» ones.
+
+```cpp
+__hash_code __code = this->_M_hash_code_tr(__k);
+size_type __bkt = _M_bucket_index(__code);
+
+if (size() > __small_size_threshold())
+  if (__node_ptr __node = _M_find_node_tr(__bkt, __k, __code))
+    return { iterator(__node), false };
+```
+
+And finally, if there is no such key in the hash table, we create a new node
+and insert it into the data structure.
+
+```cpp
+_Scoped_node __node {
+  __node_builder_t::_S_build(std::forward<_Kt>(__k),
+                 std::forward<_Arg>(__v),
+                 __node_gen),
+  this
+};
+auto __pos
+  = _M_insert_unique_node(__bkt, __code, __node._M_node);
+__node._M_node = nullptr;
+return { __pos, true };
+```
+
+Actual insertion logic is hidden inside `_M_insert_unique_node` [method][19].
+There we decide if the hash table requires a rehash and insert a node into
+the beginning of the bucket.
+
+```cpp
+  const __rehash_state& __saved_state = _M_rehash_policy._M_state();
+  std::pair<bool, std::size_t> __do_rehash
+= _M_rehash_policy._M_need_rehash(_M_bucket_count, _M_element_count,
+                  __n_elt);
+
+  if (__do_rehash.first)
+{ 
+  _M_rehash(__do_rehash.second, __saved_state);
+  __bkt = _M_bucket_index(__code);
+}
+  
+  this->_M_store_code(*__node, __code);
+    
+  // Always insert at the beginning of the bucket.
+  _M_insert_bucket_begin(__bkt, __node);
+  ++_M_element_count;
+  return iterator(__node);
+```
+
+There is a lot of interesting things inside `_M_rehash_policy._M_need_rehash`,
+but I don't want to bore you with too many details, only mention the fact that
+the number of buckets in the hash table is a [prime number][20].
+
+
+## Find
+
+`_Hashtable::find` has the same optimization for small hash tables as `insert`,
+for «small» hash tables with «slow» hash function we will do a
+[linear search first][21], otherwise do a usual [bucket search][22].
+
+```cpp
+__hash_code __code = this->_M_hash_code(__k);
+std::size_t __bkt = _M_bucket_index(__code);
+return const_iterator(_M_find_node(__bkt, __k, __code));
+```
+
+`_M_find_node` is implemented through the call to `_M_find_before_node`.
+
+```cpp
+  __node_ptr
+  _M_find_node(size_type __bkt, const key_type& __key,
+       __hash_code __c) const
+  {    
+__node_base_ptr __before_n = _M_find_before_node(__bkt, __key, __c);
+if (__before_n)
+  return static_cast<__node_ptr>(__before_n->_M_nxt);
+return nullptr;
+  }
+```
+
+`_M_find_before_node` is a good building block to have if you'll think about
+`erase` implementation as we need to remove an element from the singly linked
+lists, so having a pointer to a previous node comes in handy.
+
+`_M_find_before_node` does mostly what we expect it to do, but has a couple of
+interesting things in the sleeve.
+
+We [locate][23] a pointer to the element before the first bucket element.
+
+```cpp
+  __node_base_ptr __prev_p = _M_buckets[__bkt];
+  if (!__prev_p)
+return nullptr;
+```
```

Comment:
Is this code correct? Feels like return under if [is missed ](https://github.com/gcc-mirror/gcc/blob/b9b7981f3d6919518372daf4c7e8c40dfc58f49d/libstdc%2B%2B-v3/include/bits/hashtable.h#L817)and should we shift the `return nullptr;`?